src/core/kmod-setup.c

```diff
@@ -107,6 +107,27 @@ static bool has_virtio_console(void) {
         return r > 0;
 }
 
+static bool has_virtio_vsock(void) {
+        int r;
+
+        /* Directory traversal might be slow, hence let's do a cheap check first if it's even worth it */
+        if (detect_vm() == VIRTUALIZATION_NONE)
+                return false;
+
+        r = recurse_dir_at(
```

Comment:
BTW, why `recurse_dir_at()` is used instead of sd_device_enumerator ??

Reply:
the code is just too old i guess. and was ported over from even older code...

but also, we just care about the modalias thing, i.e. no udev db, hence I think it isn't even that bad the current way.